File: 백준/Silver/2304. 창고 다각형/창고 다각형.py
import sys

input = lambda: sys.stdin.readline().strip()

n = int(input())
arr = [0] * 1002

for i in range(n):
    x, y = map(int, input().split())
    arr[x] = y

prefix = [0] * 1002
suffix = [0] * 1002

for i in range(1, 1001):
    prefix[i] = max(arr[i], prefix[i - 1])

for i in range(1000, 0, -1):
    suffix[i] = max(arr[i], suffix[i + 1])

area = 0
for i in range(1, 1001):
    area += min(suffix[i], prefix[i])

print(area)

File: 백준/Silver/2304. 창고 다각형/창고 다각형.java
import java.util.*;
import java.io.*;

public class Main {
    public static final int size = 1002;
    static int N;
    static int[] arr = new int[size];
    static int[] prefix = new int[size + 1];
    static int[] suffix = new int[size + 1];

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;

        N = Integer.parseInt(br.readLine());

        for (int i = 0; i < N; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            arr[a] = b;
        }

        for (int i = 1; i < size; i++) {
            prefix[i] = Math.max(arr[i], prefix[i - 1]);
        }

        for (int i = size - 1; i > 0; i--) {
            suffix[i] = Math.max(suffix[i + 1], arr[i]);
        }

        int sum = 0;
        for (int i = 0; i < size; i++) {
            sum += Math.min(prefix[i], suffix[i]);
        }

        System.out.println(sum);

    }
}
